### ISSUE TITLE
Feature/Guidance Request(s): Handling unreliable connections and circumventing auto-updating

### ISSUE THREAD

**Issue by Korijn** [nwjs/nw.js#3058]
I'll just get straight to the feature requests and explain afterwards:

* Configurable error handling for network errors (HTTP errors, timeouts, proxy errors, etc.) *when using node-remote and having the entire app remotely, e.g. https://appnw.example.com/*
  * Possibly infinite retry schemes to support going offline or unreliable connections
  * Custom error pages for user friendly messages
* Load npm modules through webkit to partially circumvent need for auto-updating
* Restrict external hostnames that can be loaded (independently of the node-remote setting) to keep users in external app

The end-users of my application are often behind corporate network security systems such as firewalls and proxies, which often block and/or drop connections or even kill streams half-way through.

Access to temporary directories for auto updating is unreliable because of corporate security policies.

**tl;dr:**
* Have to handle broken connections (timeouts, packets dropped/corrupted)
* Have to handle corporate security measures (file system writing or temporary directories prohibited)

I've done a lot of research on node-webkit but I believe that handling the broken connections and security mechanisms is still hard to solve. For example, if the tool can't reach the startup page (which is external), it should display a user friendly message, explaining to them that there is likely a firewall issue or other connectivity problem. In the background, the tool should continue to try to connect and just load the page when it can.

Being able to load npm modules through webkit would be a great way to (partially) circumvent the need for auto updates.

Now, I'm unsure if any of the above can already be handled. I would love to hear your thoughts. :)

**Comment by Korijn** [comment]
Let me add that I would love to help out implementing the above if any programming work is be required. I would just prefer hearing your thoughts before I dive in.

**Comment by RIAEvangelist** [comment]
I believe that you can do all of what you desire.

One approach might be to load the external site in via a node enabled iFrame instead of a nw remote frame.

` Configurable error handling for network errors (HTTP errors, timeouts, proxy errors, etc.) when using node-remote and having the entire app remotely, e.g. https://appnw.example.com/ `
` Possibly infinite retry schemes to support going offline or unreliable connections `
Easily done if you switch to use XHR2 to request your files.

` Custom error pages for user friendly messages `
Easily done in basic javascript by handling errors when they happen and choosing what to do.

` Load npm modules through webkit to partially circumvent need for auto-updating `
you may also be able to do this with XHR2

` Restrict external hostnames that can be loaded (independently of the node-remote setting) to keep users in external app `
This would be a great example of why to use the iframe approach as the wrapping app could easily only allow access to a set list of domains etc.

None of these seem like things which can not be accomplished with the tools we have at our disposal, and I don't really know that they should be implemented at a higher level because of that.

Just my 2 cents though.

**Comment by Korijn** [comment]
Thanks for your enlightening response. I'm still wondering though, if the user clicks a link in the node enabled iframe and their connection times out. They would see a webkit error message, no? Is there some way to monitor these requests? Like an error event you can subscribe to on the iframe?

**Comment by RIAEvangelist** [comment]
I have an answer for you that will handle anything, but it is only for when shit really hits the fan.

How about you try to create the scenario you just described and share the error text here. Then I'll see if I have a less aggressive solution for you. If not I will share the crap + fan fix.

Unplug the network cable or turn of your wifi if you have to. I am interested in the exact error that pops up.

**Comment by Korijn** [comment]
Challenge accepted! ;) I'll get back to you later.

**Comment by Korijn** [comment]
I've set up Charles [1] to throttle bandwidth to 28.8 kbps and increase round-trip latency to 30 seconds. This results in a blank screen. No error message at all. Manifest:

"main": "http://some.website.com",
"node-remote": "http://some.website.com"

Using node-webkit v0.11.4 32bit (win)

[1] http://www.charlesproxy.com/

**Comment by RIAEvangelist** [comment]
Suggestion 1
#issuecomment-73182320
` One approach might be to load the external site in via a node enabled iFrame instead of a nw remote frame. `

Give that one a shot and check the dev tools.
I'm pretty sure you would be able to handle the error.

If you don't see anything helpful, post again like you did here with a screen shot of your devtools as well. Make sure you check the network pane.

**Comment by Korijn** [comment]
Well, it took some effort to get the event listeners set up right but the following code will load a page from a local webserver, endlessly retrying. I haven't tried a POST request yet, but for GET this appears to work just fine. I don't have the time to work much more on this for now so I'll post back with more results later. Thanks for the help so far.

**index.html**

    <!DOCTYPE html>
    <html>
        <head>
            <meta charset="utf-8">
            <title>Experiment</title>
            <link rel="stylesheet" href="css/app.css">
            <script src="js/app.js"></script>
        </head>
        <body>
        </body>
    </html>

**js/app.js**

    var gui = require('nw.gui');
    var rootWindow = gui.Window.get();

    var targetHost = "http://localhost:8080/";

    var statusbar = null;
    var iframe = null;
    var hideStatusBar = function() {
        if(statusbar != null)
            statusbar.style.display = 'none';
    };
    var showStatusBar = function() {
        if(statusbar != null)
            statusbar.style.display = 'block';
    };

    var firstLoadDone = false;
    var monitorStartTime = null;
    var monitorInterval = null;
    var monitorFunction = function() {
        showStatusBar();
        var current = new Date().getTime() - monitorStartTime;
        console.log("loading... " + current);
    };

    window.onload = function() {
        
        // attach event listeners
        rootWindow.on('loading', function() {
            monitorInterval = setInterval(monitorFunction, 1000);
            monitorStartTime = new Date().getTime();
        });
        
        rootWindow.on('loaded', function() {
            if(!firstLoadDone) {
                firstLoadDone = true;
            } else {
                clearInterval(monitorInterval);
                var loadSucceeded = iframe.contentWindow.location.href.indexOf("http") > -1;
                if(loadSucceeded) {
                    hideStatusBar();
                } else {
                    console.log("load failed");
                    setTimeout(function() {
                        if(iframe != null)
                            iframe.contentWindow.location.reload(true);
                    }, 2 * 1000);
                }
            }
        });
        
        // insert status bar
        statusbar = document.createElement("div");
        statusbar.setAttribute("id", "statusbar");
        statusbar.appendChild(document.createTextNode("loading"));
        document.body.appendChild(statusbar);
        
        // insert app window
        iframe = document.createElement("iframe");
        iframe.setAttribute("nwfaketop", "nwfaketop");
        iframe.setAttribute("id", "appwindow");
        document.body.appendChild(iframe);
        iframe.setAttribute("src", targetHost);
    };

**Comment by RIAEvangelist** [comment]
Nice. You could use XHR instead then you would get timeouts etc.

so make something like

    <html>
         <head>
            //stuff n scripts
        </head>
        <body>
            <iframe nwfaketop='true'>
                     //loading gui
            </iframe>
        </body>
    </html>

Then in your JS (pseudo code)

      XHR=new XMLHTTPRequest();
      XHR.timeout=1000;
      XHR.on('error',//handle it);
      XHR.on('abort',//handle it);
      XHR.on('timeout',//handle it);
      XHR.on('success',//handle it and put it in iframe);
      XHR.send('/my/page');

Thats just pseudo code but should give you the idea. I would suggest taking a look at XHR2
http://www.html5rocks.com/en/tutorials/file/xhr2/

I think this will be your best solution.

**Comment by RIAEvangelist** [comment]
also check out https://developer.mozilla.org/en-US/docs/Web/API/XMLHttpRequest
MDN may have some useful info for you on the various event types.

**Comment by Korijn** [comment]
How would I handle requests that are initiated inside of the iframe then?

**Comment by RIAEvangelist** [comment]
My first suggestion would be to have a link capture function to prevent links from leaving the page and instead use XHR to update the content.

**Comment by Korijn** [comment]
In the end I never figured out how to do this properly (in a way that isn't prone to errors) and it looks like it's simply unsupported. Too bad!
